src/Bicep.Core/TypeSystem/TypeHelper.cs

```diff
@@ -688,37 +687,42 @@ private static ObjectType TransformProperties(ObjectType input, Func<NamedTypePr
                 input.Name,
                 input.ValidationFlags,
                 input.Properties.Values.Select(transformFunc),
-                input.AdditionalProperties,
+                input.AdditionalProperties is not null ? input.AdditionalProperties with { } : null,
                 input.MethodResolver.functionOverloads);
         }
 
         public static ObjectType MakeRequiredPropertiesOptional(ObjectType input)
             => TransformProperties(input, p => p with { Flags = p.Flags & ~TypePropertyFlags.Required });
 
         public static TypeSymbol RemovePropertyFlagsRecursively(TypeSymbol type, TypePropertyFlags flagsToRemove)
-            => RemovePropertyFlagsRecursively(type, flagsToRemove, new());
+            => ModifyPropertyFlagsRecursively(type, f => f & ~flagsToRemove, new());
 
-        private static TypeSymbol RemovePropertyFlagsRecursively(
-            TypeSymbol type,
-            TypePropertyFlags flagsToRemove,
-            ConcurrentDictionary<ObjectType, ObjectType> transformedObjectCache) => type switch
+        public static ObjectType DeepCloneAndModifyPropertyFlagsRecursively(ObjectType type, Func<TypePropertyFlags, TypePropertyFlags> transformFlags) =>
```

Comment:
[nit] The `DeepClone` prefix isn't necessary because we only use immutable datastructures in this codebase.

Reply:
Will remove this method based on the other comment about immutability.